src/engine/contactwriter.cpp

```diff
@@ -4631,19 +4630,55 @@ QContactManager::Error ContactWriter::syncUpdate(const QString &syncTarget,
                 query.reportError("Failed to fetch local constituent ids for sync update");
                 return QContactManager::UnspecifiedError;
             }
+
+            QSet<quint32> allStcIds;
+            QMultiHash<quint32, quint32> modifiedToStcId;
+            QHash<quint32, quint32> aggregateForStcId;
             while (query.next()) {
                 const quint32 stConstituentId = query.value<quint32>(0);
                 const quint32 modifiedConstituentId = query.value<quint32>(1);
                 const quint32 aggregateId = query.value<quint32>(2);
-
                 constituentAggregateIds.insert(modifiedConstituentId, aggregateId);
+                if (stConstituentId) {
+                    modifiedToStcId.insertMulti(modifiedConstituentId, stConstituentId);
+                    allStcIds.insert(stConstituentId);
+                    aggregateForStcId.insert(stConstituentId, aggregateId);
+                }
+            }
 
-                if (stConstituentId && (stConstituentId != modifiedConstituentId)) {
-                    // We may need to modify the sync target constituent also
-                    stConstituents.insert(modifiedConstituentId, stConstituentId);
-                    affectedContactIds.insert(stConstituentId);
-
-                    constituentAggregateIds.insert(stConstituentId, aggregateId);
+            Q_FOREACH (quint32 modifiedId, modifiedToStcId.keys()) {
+                if (allStcIds.contains(modifiedId)) {
+                    // this modified constituent is also a sync target constituent.
+                    // we don't need to create a mapping.
+                } else {
+                    // this modified constituent is NOT a sync target constituent.
+                    // find an appropriate sync target constituent for this modified contact.
+                    bool found = false;
+                    QList<quint32> stcIds = modifiedToStcId.values(modifiedId);
+                    quint32 fallbackStc = 0;
+                    Q_FOREACH (quint32 stcId, stcIds) {
+                        if (modifiedToStcId.contains(stcId)) {
+                            // this stc was also modified.  fall back to this if no others exist.
+                            fallbackStc = stcId;
+                        } else {
+                            // unmodified stc.  Use this if possible.
```

Comment:
You could drop this stanza by setting fallbackStc to the selected ID, and dropping the 'found' variable.